Add inline emoji search

eranl opened the pull request:
Adding focused emoji suggestions for words starting with `:`. 
`~`s in the search string are converted to spaces during lookup, in order to support multiple-word search strings.
Includes all characters except white space in the emoji search string when initially typed, in order to support search strings with symbols and multiple words. 
However, when restarting a search after moving the cursor etc., only the enclosing word characters are considered, in order to keep it simple. In most cases, the search string will likely be just one word. Is this inconsistency acceptable?
Copied some dictionary support logic from #1542.
Added logic for falling back to a country-less locale when loading a dictionary, which was necessary e.g. for the *Lao (Laos)* language. Should that logic be moved somewhere else?
Made a minor change to `RichInputConnection.isCursorTouchingWord`, so that it doesn't consider emojis as part of a word, in order to support an emoji search right after an existing emoji. This also helps with regular suggestions after an emoji, which are currently missing.
The dictionary library seems to apply sophisticated language analysis, which often results in weird emoji suggestions. Is there any way to use a simpler lookup algorithm?

eranl commented:
> Repurposing the action key could work. It could change the icon to clarify the select-emoji action.

Done. Had to change the order of `InputTypeUtils.getImeOptionsActionIdFromEditorInfo`, to make this work when the action is *Enter*. 
Are there any restrictions on `actionId`? Do I need to handle the case of an existing custom action, and restore it after a search?

> Isn't there already something very similar in `DictionaryFactory`? I think it's about finding the dictionary with the _best match_ locale from assets.

Yes, but the logic for cached dicts is different. I pushed this logic down to `getCachedDictsForLocale`, since I think it could help with other dictionaries too. Let me know if that's ok.

> > Good idea. Can we use the existing autocorrect setting or do we need a separate one?
> 
> Good question. Maybe we can just have it always enabled?

I used the existing setting. When *Done* is pressed, if autocorrect is on, the default suggestion is accepted, otherwise the search string is committed.

> Made a minor change to `RichInputConnection.isCursorTouchingWord`, so that it doesn't consider emojis as part of a word, in order to support an emoji search right after an existing emoji. This also helps with regular suggestions after an emoji, which are currently missing.

I moved this change to `StringUtils.endsWithWordCodepoint` and updated the test.

`InputLogic` is very complex, and there's a good chance I didn't handle all cases. Can you recommend a set of tests to perform?

eranl commented:
Changes:
- Only do emoji search and show setting if toolbar is in compatible mode.
- Auto-show emoji suggestions.

eranl commented:
Made changes to `handleBackspaceEvent` that enable restarting suggestions when cursor is in the middle of a word. This helps with both regular suggestions and emoji ones.
Unify suggestion strip update mechanisms. Removed the `handleMessage`-based one.

eranl commented:
Replaced `isSuggestionsEnabledPerUserSettings()` with `needsToLookupSuggestions()` in several places, in order to fix autocorrect. This helps with both regular autocorrect and emoji.
Fixed `mSuggestionsEnabledPerUserSettings` value.
Moved setting.

Helium314 commented:
Did not look at the code yet, this is just from testing:
* Entering `new` shows "old style" emoji suggestions from dictionary with 🆕 first, but `:new ` doesn't even have 🆕 the first 3 (similar differences for other terms). Not sure what is causing this, but I'd expect users to be not happy with this difference. Anyway, doesn't necessarily need to be addressed in this PR.
* Emoji search mode is indicated with the _Done_ text on action key, which is also shown for the `Done` action when using Holo icon style. Did you try showing an emoji icon on the action key?

Helium314 commented on app/src/main/java/helium314/keyboard/latin/LatinIME.java:
```diff
@@ -246,13 +245,8 @@ public void handleMessage(@NonNull final Message msg) {
                             latinIme.getCurrentRecapitalizeState());
                     break;
                 case MSG_SHOW_GESTURE_PREVIEW_AND_SUGGESTION_STRIP:
-                    if (msg.arg1 == ARG1_NOT_GESTURE_INPUT) {
```
Why the changes to the UIHandler? Maybe I don't have the right overview yet, but so far I don't understand the reason.

eranl replied:
Reverted.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/SingleDictionaryFacilitator.kt:
```diff
@@ -17,6 +18,28 @@ import java.util.concurrent.TimeUnit
+     * Other suggestion fields of combined matches are taken arbitrarily from one of them.
+     */
+    fun getSuggestions(words: List<String>): SuggestionResults {
+        val infos = mutableMapOf<String, Pair<SuggestedWords.SuggestedWordInfo, Long>>()
+        words.forEach {
+            getSuggestions(it).forEach { infos.put(it.word, Pair(it, (infos[it.word]?.second ?: 0L) + it.mScore - Int.MIN_VALUE)) }
+        }
+        val suggestionResults = SuggestionResults(SuggestedWords.MAX_SUGGESTIONS, false, false)
+        suggestionResults.addAll(infos.values.map {
+            val info = it.first
+            SuggestedWords.SuggestedWordInfo(
+                info.word, info.mPrevWordsContext, (it.second / words.size + Int.MIN_VALUE).toInt(), info.mKindAndFlags,
+                info.mSourceDict, info.mIndexOfTouchPointOfSecondWord, info.mAutoCommitFirstWordConfidence
+            )
+        })
+        return suggestionResults
```
I find this rather hard to read. Could you make it more understandable? E.g.
```kotlin
        val suggestionResults = SuggestionResults(SuggestedWords.MAX_SUGGESTIONS, false, false)
        words.flatMap { getSuggestions(it) }
            .groupBy { it.word }
            .forEach { (_, results) ->
                val info = results.maxBy { it.mScore }
                val score = <insert calculation here>
                suggestionResults.add(
                    SuggestedWords.SuggestedWordInfo(
                        info.word, info.mPrevWordsContext, score, info.mKindAndFlags,
                        info.mSourceDict, info.mIndexOfTouchPointOfSecondWord, info.mAutoCommitFirstWordConfidence
                    )
                )
            }
```
(such an approach also allows using info from the highest scoring suggestion instead of using whatever comes up first)

Helium314 commented on app/src/main/java/helium314/keyboard/latin/common/StringUtils.kt:
```diff
@@ -95,19 +95,19 @@ fun getFullEmojiAtEnd(text: CharSequence): String {
-// todo: this returns true on numbers, why isn't Character.isLetter(code) used?
 fun endsWithWordCodepoint(text: String, spacingAndPunctuations: SpacingAndPunctuations): Boolean {
     if (text.isEmpty()) return false
-    var codePoint = 0 // initial value irrelevant since length is always > 0
+    var codePoint = Constants.NOT_A_CODE
     loopOverCodePointsBackwards(text) { cp, _ ->
-        codePoint = cp
-        !spacingAndPunctuations.isWordConnector(cp)
+        val bool = !spacingAndPunctuations.isWordConnector(cp)
+        if (bool)
+            codePoint = cp
+        bool
     }
-    // codePoint might still be a wordConnector (if text consists of wordConnectors)
-    return !spacingAndPunctuations.isWordConnector(codePoint) && !spacingAndPunctuations.isWordSeparator(codePoint)
+    return codePoint != Constants.NOT_A_CODE && spacingAndPunctuations.isWordCodePoint(codePoint)
```
Can you explain your reasoning? The changes don't appear to be related to emoji search.

eranl replied:
As I explained [here](https://github.com/Helium314/HeliBoard/pull/1660#issuecomment-2960778782):


>>    Made a minor change to `RichInputConnection.isCursorTouchingWord`, so that it doesn't consider emojis as part of a word, in order to support an emoji search right after an existing emoji. This also helps with regular suggestions after an emoji, which are currently missing.

> I moved this change to `StringUtils.endsWithWordCodepoint` and updated the test.

Helium314 replied:
Sound like a reasonable change. Though I'd prefer to have such general improvements done separately.

eranl replied:
Should I create a separate PR then?

eranl replied:
Done - #1814.

Helium314 replied:
Thanks!

eranl commented:
> > Are there any restrictions on actionId?
> 
> Judging by [IME_MASK_ACTION](https://developer.android.com/reference/android/view/inputmethod/EditorInfo#IME_MASK_ACTION) it should be in the range from 0-255.

Is there a risk of collision with an app-provided `actionId`?

> > Made changes to handleBackspaceEvent that enable restarting suggestions when cursor is in the middle of a word. This helps with both regular suggestions and emoji ones.
> 
> Did you test something here?

Not sure what you mean. It seems to work fine to me.

> 
> > Replaced calls to isSuggestionsEnabledPerUserSettings() with needsToLookupSuggestions() in several places, in order to fix autocorrect. This helps with both regular autocorrect and emoji.
> 
> How does it help?

This makes autocorrect work when suggestions are off.

> > Fixed mSuggestionsEnabledPerUserSettings value.
> 
> What did you fix?

The logic was wrong. I fixed the case of `suggestionsEnabled` off and `mOverrideShowingSuggestions` on.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -406,6 +416,11 @@ public boolean moveCursorByAndReturnIfInsideComposingWord(int distance) {
         return mWordComposer.moveCursorByAndReturnIfInsideComposingWord(distance);
     }
 
+    public void onSettingsChanged(Locale locale) {
```
I don't really like this, because there is a slight but existing difference between `loadSettings` and `onSettingsChanged`, and because I fear this is sort of inviting for future changes to do more and more in the future.

Better just call `updateEmojiDictionary` directly from `loadSettings`, and maybe name it `updateEmojiDictionaryIfNecessary` (similar to `resetDictionaryFacilitatorIfNecessary`).

eranl replied:
Doesn't this change reduce coupling between `InputLogic` and `LatinIME`? Isn't it better if `LatinIME` (or any other caller) doesn't know the implementation details of what  `InputLogic` does when settings change?

Helium314 replied:
I see your point, and if `InputLogic` wasn't the monstrosity it currently is, then it might just be fine.
`InputLogic` is doing too much already, my experience is that contributors like to add stuff here because "it fits good enough", making the class even harder to manage (same thing goes for LatinIME).

Having a generic `onSettingsChanged` is a very inviting place to put more and more functionality, and it's not going to make the (at some point inevitable) complete overhaul any simpler.
So in the end it's not your current changes I fear, but what might come (would be different if I actually could find the time to redo `InputLogic` anytime soon).

eranl replied:
Done.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -1767,8 +1797,12 @@ private void restartSuggestions(final TextRange range) {
             // if shouldIncludeResumedWordInSuggestions is true, 0 otherwise. In this case, we
             // have no useful suggestions, so we will try to compute some for it instead.
+            final AsyncResultHolder<SuggestedWords> holder = new AsyncResultHolder<>("Suggest");
             mInputLogicHandler.getSuggestedWords(() -> getSuggestedWords(Suggest.SESSION_ID_TYPING,
-                SuggestedWords.NOT_A_SEQUENCE_NUMBER, this::doShowSuggestionsAndClearAutoCorrectionIndicator));
+                                                                         SuggestedWords.NOT_A_SEQUENCE_NUMBER, holder::set));
+            // This line may cause the current thread to wait.
+            final SuggestedWords suggestedWords = holder.get(null, Constants.GET_SUGGESTED_WORDS_TIMEOUT);
+            doShowSuggestionsAndClearAutoCorrectionIndicator(suggestedWords);
```
Why this change, and why make the thread wait?

eranl replied:
Looks like this was a mistake. Will revert, but it's pending the decision on [this comment](https://github.com/Helium314/HeliBoard/pull/1660#discussion_r2173840544).

eranl replied:
Reverted.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -1782,9 +1816,19 @@ private void restartSuggestions(final TextRange range) {
+        }
+    }
+
+    private boolean shouldAutoShowSuggestions(SettingsValues settingsValues, SuggestedWords suggestedWords) {
+        return ! suggestedWords.isEmpty() && settingsValues.isSuggestionsEnabledPerUserSettings() && ! mWordComposer.isResumed()
+                    && isEmojiSearch();
```
This is confusing. Why should suggestions be shown automatically only if it's an emoji search?

Helium314 replied:
I think what had me confused is the name. `shouldAutoShowSuggestions` seems to be used to determine whether the suggestion _strip_ is shown.
Since it's used the same way in both uses, maybe the check and action could be merged into some `showSuggestionStripIfEmojiSearch`.

Helium314 commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
I find the changes here pretty hard to review, mostly because it's many small changes scattered over this already overly complex class (meaning I'll have to go through them again at least once I guess...).
The addition of emoji search related code "everywhere" gives it a hacky feeling, but currently it's not more than a feeling.

Anyway, could you at least group the emoji search related methods, so they are in a continuous place?

eranl replied:
> Anyway, could you at least group the emoji search related methods, so they are in a continuous place?

Done.

eranl commented on app/src/main/java/helium314/keyboard/latin/settings/SettingsValues.java:
```diff
@@ -237,8 +239,8 @@ public SettingsValues(final Context context, final SharedPreferences prefs, fina
                 && ((inputAttributes.mInputType & InputType.TYPE_MASK_VARIATION) != InputType.TYPE_TEXT_VARIATION_WEB_EDIT_TEXT
                   || !prefs.getBoolean(Settings.PREF_ALWAYS_SHOW_SUGGESTIONS_EXCEPT_WEB_TEXT, Defaults.PREF_ALWAYS_SHOW_SUGGESTIONS_EXCEPT_WEB_TEXT));
         final boolean suggestionsEnabled = prefs.getBoolean(Settings.PREF_SHOW_SUGGESTIONS, Defaults.PREF_SHOW_SUGGESTIONS);
-        mSuggestionsEnabledPerUserSettings = ((mInputAttributes.mShouldShowSuggestions && suggestionsEnabled)
-                || mOverrideShowingSuggestions) && !mSuggestionStripHiddenPerUserSettings;
+        mSuggestionsEnabledPerUserSettings = suggestionsEnabled && (mInputAttributes.mShouldShowSuggestions || mOverrideShowingSuggestions)
+                && !mSuggestionStripHiddenPerUserSettings;
```
You're right - it's not specific to emoji search. Will do.

eranl commented on app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripViewAccessor.java:
```diff
@@ -13,5 +13,6 @@
 public interface SuggestionStripViewAccessor {
     void setNeutralSuggestionStrip();
-    void showSuggestionStrip(final SuggestedWords suggestedWords);
+    void setSuggestions(final SuggestedWords suggestedWords);
+    void showSuggestionStrip();
```
The name `showSuggestionStrip` was misleading, because it doesn't really show the strip - it only sets the suggestions. So I renamed it, and added a new true `showSuggestionStrip` method.

Helium314 replied:
Makes sense, thanks!

eranl commented:
> * Entering `new` shows "old style" emoji suggestions from dictionary with 🆕 first, but `:new ` doesn't even have 🆕 the first 3

Thanks for the catch! This was caused by the dictionary returning duplicate results + `Int` overflow. Fixed.

> * Emoji search mode is indicated with the _Done_ text on action key, which is also shown for the `Done` action when using Holo icon style. Did you try showing an emoji icon on the action key?

Changed it to 👍. Is that what you had in mind? I tried to change it to the standard Android *Done* action character, but couldn't find it.

eranl commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -2416,6 +2490,38 @@ public void getSuggestedWords(final int inputStyle, final int sequenceNumber, fi
+            callback.onGetSuggestedWords(SuggestedWords.getEmptyInstance());
+        } else {
+            var input = mWordComposer.getTypedWord().substring(1);
+            var words = WHITESPACE_PATTERN.matcher(input).replaceAll(" ").split(" ");
```
Thanks for reminding me about it. Done.

eranl commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -2208,6 +2266,19 @@ public void onUpdateTailBatchInputCompleted(final SettingsValues settingsValues,
+            getCurrentInputEditorInfo().actionId = on? EMOJI_SEARCH_DONE_ACTION : 0;
+            KeyboardSwitcher.getInstance().reloadKeyboard();
+        }
+    }
```
> * `setEmojiSearch` sets an actionId to `EditorInfo`, but `isEmojiSearch` checks something completely different. That should at least be clarified and reasoned in a comment

Renamed it to `setEmojiSearchAction`.

> * I have doubts that changing the `EditorInfo` will be working completely without side effects. Maybe there is a better way of influencing `TextKeyData.getActionKeyLabel`? Getting the emoji search state would then need to be in `SettingsValues` I guess.

I agree. I actually think now that it's wrong for the IME to set these fields. How about storing an `InternalAction(id, label, description)` on `Keyboard`?

eranl replied:
Moved the custom action to `KeyboardId`.

eranl commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -1022,6 +1040,11 @@ private void handleNonSeparatorEvent(final Event event, final SettingsValues set
     }
 
+    private boolean isEmojiSearch(int codePoint) {
+        return mEmojiDictionaryFacilitator != null && (! mWordComposer.isComposingWord() && codePoint == ':'
```
Done.

eranl commented:
Should I rename it in code everywhere from `EmojiSearch` to `EmojiSuggestions`, to differentiate from the upcoming true emoji search?

eranl commented:
> > Should I rename it in code everywhere from EmojiSearch to EmojiSuggestions, to differentiate from the upcoming true emoji search?
> 
> Good idea!

Sorry about the back and forth, but EmojiSuggestions might also be confusing, since we have emoji suggestions outside of this feature too. I used the term InlineEmojiSearch in one place, and *Focused emoji suggestions* in the UI. Which one do you think is best?

Helium314 commented:
> Which one do you think is best?

I would prefer _InlineEmojiSearch_, because it fits a little better. Even though technically it's mostly using suggestion-related parts of the app, I guess a user would rather call it _search_.

eranl commented:
> I would prefer _InlineEmojiSearch_

Done.

Helium314 commented:
Would it be much work to only enter emoji search mode after typing a character after the `:` (and maybe don't enter search mode if it's space)?
This could reduce (presumably) unwanted activations of the search mode, but I'm not sure a) whether this is really true, and b) whether it's worth the effort.

Helium314 commented:
I have another look over the changes to `InputLogic` (trying to find potential issues), but I hope it's finally time to merge.

eranl commented on app/src/main/java/helium314/keyboard/keyboard/internal/keyboard_parser/floris/KeyCode.kt:
```diff
@@ -174,6 +174,7 @@ object KeyCode {
     const val ALT_RIGHT =                 -10047
     const val META_LEFT =                 -10048
     const val META_RIGHT =                -10049
+    const val INLINE_EMOJI_SEARCH_DONE =  -10051
```
Because I used -10050 in #1765 🙂

eranl commented on app/src/test/java/helium314/keyboard/latin/InputLogicTest.kt:
```diff
@@ -97,10 +97,7 @@ class InputLogicTest {
-        //  setting whole word composing will delete text behind cursor
-        //  setting part before cursor as composing may be bad if user just wants to adjust a letter and result is some autocorrect
-        assertEquals("", composingText)
+        assertEquals("yu", composingText)
```
Done. #1824.

eranl commented:
> When enabling the setting, there could be a warning dialog if there is no emoji dictionary added (either none at all, or none for the current locale).

Done. Should we add this to emoji descriptions too, replacing the summary?

Helium314 commented:
There seems to be some issue with glide typing. When glide typing there is a space inserted between `:` and the word. This can be disabled using the _autospace before word_ setting, but maybe this autospace should be disabled during inline emoji search.
Anyway, when disabling this setting, glide typing can be used to write after `:`, but seems to exit inline emoji search.

> Replaced calls to `isSuggestionsEnabledPerUserSettings()` with `needsToLookupSuggestions()` in several places, in order to fix autocorrect. This helps with both regular autocorrect and emoji.

Oh, actually this is another thing that might be better of in a separate PR. But it's minor enough that it's not worth the work of extracting.

> Should we add this to emoji descriptions too, replacing the summary?

Good idea, definitely makes sense.

eranl commented on app/src/main/java/helium314/keyboard/latin/inputlogic/InputLogic.java:
```diff
@@ -791,6 +798,17 @@ private void handleFunctionalEvent(final Event event, final InputTransaction inp
+                    commitCurrentAutoCorrection(Settings.getValues(), LastComposedWord.NOT_A_SEPARATOR, handler);
+                    inputTransaction.setDidAutoCorrect();
+                } else {
+                    commitTyped(Settings.getValues(), LastComposedWord.NOT_A_SEPARATOR);
```
Done.

eranl commented:
> There seems to be some issue with glide typing. When glide typing there is a space inserted between `:` and the word. This can be disabled using the _autospace before word_ setting, but maybe this autospace should be disabled during inline emoji search. Anyway, when disabling this setting, glide typing can be used to write after `:`, but seems to exit inline emoji search.

This is quite a big miss on my part... In fact, I'm struggling to figure out how to make this work with glide typing, since it requires making emoji suggestions on the result of the glide suggestions... Can you please offer some guidance?

> > Should we add this to emoji descriptions too, replacing the summary?
> 
> Good idea, definitely makes sense.

Should I add it to this PR or as a follow-up?

> (sorry for another round of nitpicking, if it wasn't for the glide typing issue I'd just do the changes and merge)

No problem! I actually appreciate the fact that you take the time to review changes carefully and give thoughtful comments, and that you give me a chance to make the required changes myself.

eranl commented:
I can't figure out why these tests are failing. They're not failing locally.

eranl commented:
> just put it in here.

Done.

> ```java
>                 searchForEmojiInline(sequenceNumber, ???, callback);
> ```

The biggest questions for me are:
1. should the user be given a chance to pick a correction to the glide-typed word first, before suggesting emojis?
2. how to support multiple glide-typed words during inline emoji search?

Helium314 commented:
> how to support multiple glide-typed words during inline emoji search?

That's a good question. I guess that makes my approach useless.
So we may need to gather all the composed data from the different words. Or gather the suggestions for each word separately.

eranl commented:
What are the statistics of glide typing? How often does a user have to pick a correction? 

How many users use glide typing? Would it be acceptable to not support it for the initial inline emoji search release?

> Oh, and the whole glide typing process may shows suggestions during input, maybe that should also be emojis?

That's an interesting idea. Something like:
1. get the top four (?) glide typing results
2. search emojis for each one and multiply the word score by the emoji score
3. merge them

It gets complicated with multiple words though - we'd need to try all word combinations.

Also, with multiple words, how do we keep track of the inline emoji search state? Since each word gets committed separately, I can't think of any natural way to do that, so we might need to introduce an external state flag, which carries the risk of inconsistencies.

eranl commented:
> The multiple word thing could be an issue. Also I'm not sure whether users would want an emoji preview, or a word. Maybe for the first word it could be an emoji preview, and then just the word? But anyway, this is something we can postpone / add later (and only if users explicitly want it).

What do you suggest we do for now? For each word, use just the top glide suggestion (the one in composed data), ignoring the rest, and suggest emojis based on that?

> But with this I understand your concern about need for a flag and potential inconsistencies.

Any alternative?

> Anyway, since this PR will take a bit more work I'll remove it from my plans for 3.3. 

Makes sense.

smalltheif commented:
> I think it's uncommon for people to type a colon after a space in normal typing.

Maybe I misunderstand your statement and just for a quick indication, in French-language typography, it's the [rule to add a space before a colon](https://en.wikipedia.org/wiki/Colon_(punctuation)#Spacing_and_parentheses).

smalltheif commented:
> Do you think that a colon following a space is not a good enough marker for starting an inline emoji search then? Should it be space, colon, and then a non-space, as @Helium314 suggested?

Yes I think it would be preferable to start the inline emoji search with the marker _space, colon, and then a non-space_ compared to just _space, colon_.
However it seems that some people like to [get emoji suggestions without having to type a colon](https://github.com/Helium314/HeliBoard/issues/1826#issuecomment-3143958066) so if it possible we should aim for inline emoji search with _space, colon, and then a non-space_ and just _space_.

eranl commented:
> It gets complicated with multiple words though - we'd need to try all word combinations.

I tried this approach, by keeping track of all glide word suggestions and translating the glide suggestion scores to weights such that they sum up to one per glide-typed word. However, I then realized that:
- glide suggestion scores are too similar, resulting in noise
- glide and regular typing can be mixed within a single emoji search

> we'd need to ask `mConnection` to get the committed text after the last `:`. 

So I ended up using this approach, abandoning the `WordComposer` approach, relying on `mConnection.getTextBeforeCursor` instead. I used the internal action as the mode marker.

> Yes I think it would be preferable to start the inline emoji search with the marker _space, colon, and then a non-space_ compared to just _space, colon_. 

Done, with a minor twist - as before, the character before the colon can be any non-word character,

> However it seems that some people like to [get emoji suggestions without having to type a colon](https://github.com/Helium314/HeliBoard/issues/1826#issuecomment-3143958066)

This is unrelated - it's about emoji suggestions in regular typing, which is already supported.

> maybe this autospace should be disabled during inline emoji search. 

This is a catch - we can't tell if the user intended to type a normal word after the colon, or whether they intended to start an emoji search. I guess we can add a setting for this, or rely on `SpacingAndPunctuations.isUsuallyPrecededBySpace(':')`.

One issue of concern: if you have autocorrect on, there's no way to exit emoji search mode without accepting a suggestion or changing the text around the colon.

Helium314 commented:
@eranl when normally entering emojis this works really nice!
What I noticed:
* In emoji search mode, empty suggestions (or punctuation suggestions if enabled) are shown after pressing space, until the next input. When pressing the action key now, the emoji on the action key is gone, but otherwise nothing happens.
* If there is text directly after a `:`, moving the cursor after this always seems to enable emoji search mode (which could easily be unwanted), but the action key does nothing. When selecting an emoji, all text since the emoji search start is replaced, so this part seems to work. This might be related to #1892?

eranl commented:
> Well, we can just leave it and see whether this matches user expectations. Then I'd rather put it in the _experimental_ section for maybe 2 or 3 releases though.

Makes sense.

> If you can reasonably move logic so unit tests can be used, I'd very much recommend to do so. 

Let me know if this is what you had in mind.

> ⏹️ seems good to me.

Done. On my device it's orange, BTW.

> in autocorrect mode, this now means that there's no easy way out of it if unintended.
> 
> I can see a few possible solutions:
>     1. disable autocorrect for emoji search
>     2. disable autocorrect for emoji search when `mWordComposer.isResumed()`
>     3. limit emoji search triggering on input reset to some number of characters/words between the start sequence and the cursor

What's your take on this?

Helium314 commented:
> Let me know if this is what you had in mind.

Yes, this is almost exactly what I had thought of.

> What's your take on this?

What's your opinion?
I think 3. would make sense. Like using the last 50 instead of 1024 characters. I would assume emoji search would not be longer than 20 or 30 characters, so I guess 50 should be fine.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/KeyboardLayoutSet.java:
```diff
@@ -70,6 +70,8 @@ public KeyboardLayoutSetException(final Throwable cause, final KeyboardId keyboa
         }
     }
 
+    public record InternalAction(int code, String label) {}
```
Could you add some description to make clear what it does?
E.g. _key code and label overriding the action provided by the input field_.

eranl commented on app/src/main/res/values/strings.xml:
```diff
@@ -21,10 +21,14 @@
+    <!-- Description for option to perform inline emoji search -->
+    <string name="inline_emoji_search_summary">Suggest only emojis when a word is preceded by a colon</string>
+    <!-- Warning text for when enabling inline emoji search with no emoji dictionary -->
+    <string name="emoji_dictionary_required">An emoji dictionary is required for this feature to work</string>
```
Done.

eranl commented:
> What's your opinion? I think 3. would make sense. Like using the last 50 instead of 1024 characters.

Agreed. Done.

Helium314 commented on app/src/main/res/values/strings.xml:
```diff
@@ -26,7 +26,7 @@
     <string name="inline_emoji_search_summary">Suggest only emojis when a word is preceded by a colon</string>
     <!-- Warning text for when enabling inline emoji search with no emoji dictionary -->
-    <string name="emoji_dictionary_required">An emoji dictionary is required for this feature to work</string>
+    <string name="emoji_dictionary_required">An emoji dictionary is required for this feature to work. Dictionaries in .dict format can be downloaded %s.</string>
```
The string is added by this PR, so no need.
Could you add to the comment that `%s` will be replaced by `dictionary_link_text`?

Helium314 commented on app/src/main/java/helium314/keyboard/settings/screens/TextCorrectionScreen.kt:
```diff
@@ -266,6 +265,23 @@ fun createCorrectionSettings(context: Context) = listOf(
 )
 
+@Composable
+fun SwitchPreferenceWithEmojiDictWarning(setting: Setting, default: Boolean, context: Context) {
```
I'd say move it into the `switchPreference` file, but your choice.

eranl commented:
I'm not getting the build failure. It builds fine locally.
